### DISCUSSION TITLE
Connection to postgresql database problem

### DISCUSSION THREAD

**Discussion by Estegar** [usebruno/bruno#1408] [Q&A]
Hi,
i want to connect to database using:
```
require('pg')
```
so i did a setup, added this:
```
"scripts": {
    "moduleWhitelist": [
      "os", 
      "events", 
      "crypto", 
      "dns", 
      "net", 
      "assert"
    ],
    "filesystemAccess": {
      "allow": true
    }
  }
```
to my bruno.json 

i also needed to add this line:
```
const TextEncoder = require('util').TextEncoder
```
to utils-webcrypto.js
and this line:
```
var URL = require('url').URL
```
to pg-connection-string/index.js because i got TextEncoder is not defined or URL is not defined...

But i still cannot run query correctly. I get error: 
"Error invoking remote method 'send-http-request': TypeError: 'set' on proxy: trap returned falsish for property 'connectionParameters'"

 What can i do? I need to run some scripts on database before and after some of my tests.

**Comment by Estegar** [comment]
Hello, i still have this issue, could someone help me out?

**Reply by ParisseNicolas** [reply]
Hello,

Did you find a solution to execute query on a DB from Bruno?

**Reply by ParisseNicolas** [reply]
I found a solution:
```
cd location/to/your/brunoCollectionFolder
npm init -y
npm install child_process
```

**`bruno.json`**
```
{
  "version": "1",
  "name": "myAPI",
  "type": "collection",
  "ignore": [
    "node_modules",
    ".git"
  ],
  "scripts": {
    "moduleWhitelist": ["child_process"],
    "filesystemAccess": {
      "allow": true
    }
  }
}
```

**`yourRequest.bru`**
```
script:pre-request {
  const { exec } = require('child_process');
  const path = require('path');
  
  const scriptPath = path.resolve(__dirname, './assertDBConnection.js');
  
  exec(`node ${scriptPath}`, (error, stdout, stderr) => {
    if (error) {
	    console.error(`Error executing script: ${error.message}`);
	    return;
    }
    if (stderr) {
	    console.error(`Script stderr: ${stderr}`);
	    return;
    }
    console.log(`Script stdout: ${stdout}`);
  });
  }
```

**`assertDBConnection.js`**
```
const { Client } = require('pg');

const connectionString = 'postgresql://username:pwd@ip:port/DBName';

const assertDbConnection = async (connectionString) => {
    console.log(`connectionString: ${connectionString}`);
    
    const client = new Client({
        connectionString: connectionString,
    });
    
    try {
        await client.connect();
        console.log(`Successfully connected to database: ${connectionString}`);
    } catch (err) {
        console.error(`Error connecting to database: ${connectionString}`, err);
    } finally {
        await client.end();
    }
};

const run = async () => {
    await assertDbConnection(connectionString);
};

run();
```

**`Repo tree`**
```
repository root
│-- package.json   
│-- Bruno
      |-- YourCollection
           |-- assertDBConnection.js
           |-- package.json
           |-- bruno.json
           |-- aBrunoFolder
                |-- YourRequest.bru
```
> [!NOTE]
> even if path.resolve(__dirname, './assertDBConnection.js') is in YourRequest.bru it is actually located at the root of your collection hence the ./assertDBConnection.js

As you notice I have two package.json. This is because `pg` is installed on the repository root folder. If installing `pg` on the package.json of bruno it won't work because of bruno application unable to work with it properly. 
If someone find a way to avoid having the package.json in Bruno or directly target the one in the repository root would be appreciable.

**Comment by hschletz** [comment]
Many modules assume the TextEncoder object to be available by default, which is the case in recent Node versions, but not in Bruno (Bug?  By design?)

Workaround: import the object into the global scope before requiring any module that depends on it:
```js
global.TextEncoder = require('util').TextEncoder
```
The _util_ module is shipped with Bruno. It does not need to be added via npm.

This should also work with other symbols that are missing in the Bruno environment. The key is to put those symbols in the global scope. Importing into the current script would work for Bruno scripts only, but not for third party modules.

**Reply by ParisseNicolas** [reply]
Doing this just makes Bruno UI to crash instantly when reaching this statement. Weirdly, everything works fine in CLI mode.
Tested with Bruno version 2.13.2 and @usebruno\cli version 2.13.2
